### DISCUSSION TITLE
How to use array with `connectOrCreate` with explicit many-to-many

### DISCUSSION THREAD

**Discussion by mknaf** [prisma/prisma#18505] [Q&A] [answered]
Hi, I'm using an explicit many-to-many relation with a custom column ( `quantity`) on the linking table. I'm having trouble using `connectOrCreate` for more than one entry at a time.

This is my (reduced) database schema:
```prisma
model PCB {
  id Int @default(autoincrement()) @id
  name String @unique
  resistors ResistorsOnPCBs[]
}

model Resistor {
  id Int @default(autoincrement()) @id
  value Int @unique
  pcbs ResistorsOnPCBs[]
}

model ResistorsOnPCBs {
  pcb PCB @relation(fields: [pcbId], references: [id])
  pcbId Int

  resistor Resistor @relation(fields: [resistorId], references: [id])
  resistorId Int

  quantity Int @default(1)

  @@id([pcbId, resistorId])
}
```

`create`ing a `PCB` with only a single `Resistor` works. This example uses hardcoded dummy values:
```javascript
this.prisma.pCB.create({
  data: {
    name: pcb.name,
    resistors: {
      create: {
        quantity: 10,
        resistor: {
          connectOrCreate: {
            where: {
              value: 100,
            },
            create: {
              value: 100,
            },
          },
        },
      },
    },
  },
});
```

However, `create`ing a `PCB` with multiple `Resistor`s at a time gives me the error displayed below.
```javascript
this.prisma.pCB.create({
  data: {
    name: pcb.name,
    resistors: {
      create: pcb.resistors.map((resistor) => ({
        quantity: resistor.quantity,
        resistor: {
          connectOrCreate: {
            where: {
              value: resistor.value,
            },
            create: {
              value: resistor.value,
            },
          },
        },
      })),
    },
  },
});
```

```
Error: 
Invalid `prisma.pCB.create()` invocation:

{
  data: {
    name: 'test123',
    resistors: {
      create: {
        '0': {
?         quantity?: Int,
          resistor: {
          ~~~~~~~~
            connectOrCreate: {
              where: {
                value: '12345123'
              },
              create: {
                value: '12345123'
              }
            }
          },
+         resistorId: Int
        }
      }
    }
  }
}

Unknown arg `resistor` in data.resistors.create.0.resistor for type ResistorsOnPCBsUncheckedCreateWithoutPcbInput. Did you mean `resistorId`?
Argument resistorId for data.resistors.create.0.resistorId is missing.

Note: Lines with + are required, lines with ? are optional.
```

I've seen contradicting info on this now. So here are my questions:

- Is it possible to do the `create` with an array?
- Why is prisma asking for a `resistorId` when attempting a `create` with an array, but it is fine without one when doing a `create` with a single `Resistor`?

I think this may be related to https://github.com/prisma/prisma/issues/3266

**Comment by nurul3101** [comment]
Hello @mknaf 👋 

An array can be passed in a nested [createMany](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#createmany) call, not in a nested create query.

I would recommend invoking a create query inside a for loop to create multiple PCB records.

**Reply by nurul3101** [reply]
@mknaf If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by mknaf** [reply, marked as answer]
Hello @nurul3101 , thanks for answering. 

I have it solved like this now:

```ts
    // manually create or get existing resistors
    const resistors: Resistor[] = (
      await Promise.all(
        pcb.resistors.map(async (resistor) => {
          return await this.resistorService.createOrGetByValue(resistor.value);
        })
      )
    ).flat();

    // create the new pcb
    const newpcb: PCB = await this.pcbService.create({
      name: pcb.name,
      price: pcb.price,
      quantity: pcb.quantity,
      resistors: {
        create: resistors.map((res, i) => ({
          quantity: pcb.resistors[i].quantity,
          resistor: {
            connect: {
              id: res.id,
            },
          },
        })),
      },
    });
```

In the `resistorService` I use an `upsert` to either create or get a potentially existing Resistor:
```ts
  createOrGetByValue(
    value: number,
    quantity?: number,
    price?: number
  ): Promise<ResistorModel> {
    return this.prisma.resistor.upsert({
      where: {
        value,
      },
      update: {},
      create: {
        value,
        quantity,
        price,
      },
    });
  }
```

In the `pcbService` I simply pass the prepared PCB object as the `data` value:
```ts
  async create(data: Prisma.PCBCreateInput): Promise<PCB> {
    return this.prisma.pCB.create({ data });
  }
```

While this is working, I find it a bit tedious. In this particular case, there are a lot more parts on a `PCB` than just `Resistor`s, so I have to do this for all sorts of entities.

Is there a technical reason it is not working as I attempted in my original question or is it simply not implemented?

Thanks

**Reply by nurul3101** [reply]
Your solution would be an ideal way to achieve this.
